sql: Integrate RLS Enabled Field

spilchen opened the pull request:
Previously, the new row-level security (RLS) logic for reads was only functional within unit tests. This change utilizes the RLS enabled field in the table descriptor to ensure USING expressions in policies are correctly applied outside of unit tests. Additional logic tests have been added to validate this behavior.

Note: There are existing issues with the statement cache that can cause incorrect policy application for users/roles. These will be addressed in a follow-up issue.

While creating the test, I encountered two issues with dependency checking:
- `DROP CASCADE` did not correctly handle the new policy expression elements, which surfaced when dropping a database.
- Sequence back-reference maintenance was incorrect when adding a second reference that only targeted the table (as occurs when a sequence is referenced in a policy expression).

Epic: CRDB-11724
Release note: None
Informs #136717

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140028)
<!-- Reviewable:end -->

rafiss reviewed (approved):
nice work on the testing! was there already an existing test for the `drop ... cascade` behavior?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140028#-:-OHmiscu9T-UeL8t9oBQ:b-n7fdj1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

spilchen committed:
sql: Integrate RLS Enabled Field

Previously, the new row-level security (RLS) logic for reads was only
functional within unit tests. This change utilizes the RLS enabled field
in the table descriptor to ensure USING expressions in policies are
correctly applied outside of unit tests. Additional logic tests have
been added to validate this behavior.

Note: There are existing issues with the statement cache that can cause
incorrect policy application for users/roles. These will be addressed in
a follow-up issue.

While creating the test, I encountered two issues with dependency
checking:
- `DROP CASCADE` did not correctly handle the new policy expression
  elements, which surfaced when dropping a database.
- Sequence back-reference maintenance was incorrect when adding a
  second reference that only targeted the table (as occurs when a
  sequence is referenced in a policy expression).

Epic: CRDB-11724
Release note: None
Informs #136717

spilchen reviewed (commented):
Good point about the `DROP ... CASCADE` test. I encountered that error while developing the RLS test. It occurred when the logic test terminated abruptly and cleaned up by dropping the database. The test failure was due to an unrelated error, and once I fixed that, we no longer triggered the cascade code in the same way. To ensure proper coverage, I added a separate test specifically for the `DROP ... CASCADE` scenario.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140028#-:-OHnD-6R5bYcOW1DCgEz:bmzo527)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
